### ISSUE TITLE
intermittent failure: twisted.conch.test.test_conch.OpenSSHClientTestCase.test_remoteToLocalForwarding

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3193]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3193|
|Type|defect|
|Created|2008-04-18 18:52:34Z|
|Branch|https://github.com/twisted/twisted/tree/test-conch-fix-3193|

```
===============================================================================
[FAIL]: twisted.conch.test.test_conch.OpenSSHClientTestCase.test_remoteToLocalForwarding

Traceback (most recent call last):
Failure: twisted.trial.unittest.FailTest: not equal:
a = ''
b = 'test\n'

-------------------------------------------------------------------------------
```

<details><summary>Searchable metadata</summary>

```
trac-id__3193 3193
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_test_conch_fix_3193 branches/test-conch-fix-3193
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1208544754000000 1208544754000000
changetime__1209976505000000 1209976505000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Putting the failure in the description instead of a buildbot link because sometimes buildbot links go away.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

The relevant failing build was http://buildbot.twistedmatrix.com/builders/osx10.5-py2.4-select/builds/770 in case it hasn't gone away yet and somebody wants to look at the logs.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [23431]) Branching to 'test-conch-fix-3193'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Ok the fix is in the branch. Basically, I made _getFreePort returns stopListening so that the port is really free (remember, stopListening isn't synchronous). Before, the process tries to connect too quickly, manage to connect to the server we launch in _getFreePort, which closed the connection immediately.

I also changed a timing so that it tries sooner to connect to the external server (0.1 instead of 1). That speeds the tests by a fair amount.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

I made a small modification: I used a socket instead of listenTCP, so that it could happend synchronously. Hopefully the problem is still solved :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Looks great, please merge

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [23518]) Merge test-conch-fix-3193

Author: therve
Reviewer: exarkun
Fixes [#3193](https://github.com/twisted/twisted/issues/3193)

Change test_conch's method to get free listening port to be synchronous: that
removes some intermittent failures under OS X.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
